File: apps/ios/CodeBuddyRemote/CodeBuddyRemote/RelayE2E.swift
import CryptoKit
import Foundation
import Security

enum RelayE2EError: Error {
  case sessionNotEstablished
  case invalidPublicKey
  case invalidEnvelope
  case invalidCiphertext
}

final class RelayE2EPeer {
  enum Role {
    case host
    case client
  }

  let role: Role
  let publicKey: String
  let version = 1
  let algorithm = "P256-HKDF-SHA256-CHACHA20-POLY1305"

  private let pairingCode: String
  private let privateKey = P256.KeyAgreement.PrivateKey()
  private var session: RelayE2ESession?
  private var outgoingSeq = 0

  init(role: Role, pairingCode: String) {
    self.role = role
    self.pairingCode = pairingCode.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
    self.publicKey = privateKey.publicKey.x963Representation.base64URLEncodedString()
  }

  func deriveSession(peerPublicKey: String) throws {
    guard let publicKeyData = Data(base64URLEncoded: peerPublicKey) else {
      throw RelayE2EError.invalidPublicKey
    }
    let peerKey = try P256.KeyAgreement.PublicKey(x963Representation: publicKeyData)
    let sharedSecret = try privateKey.sharedSecretFromKeyAgreement(with: peerKey)
    let salt = Data("codebuddy-remote relay e2e v1\n\(pairingCode)".utf8)
    session = RelayE2ESession(
      hostToClient: sharedSecret.hkdfDerivedSymmetricKey(
        using: SHA256.self,
        salt: salt,
        sharedInfo: Data("host-to-client".utf8),
        outputByteCount: 32
      ),
      clientToHost: sharedSecret.hkdfDerivedSymmetricKey(
        using: SHA256.self,
        salt: salt,
        sharedInfo: Data("client-to-host".utf8),
        outputByteCount: 32
      )
    )
  }

  func encryptPayload(_ payload: [String: Any]) throws -> [String: Any] {
    guard let session else { throw RelayE2EError.sessionNotEstablished }
    outgoingSeq += 1
    let direction = outgoingDirection
    let plaintext = try JSONSerialization.data(withJSONObject: payload)
    let nonce = try ChaChaPoly.Nonce(data: randomBytes(count: 12))
    let sealed = try ChaChaPoly.seal(
      plaintext,
      using: session.key(for: direction),
      nonce: nonce,
      authenticating: aad(direction: direction, seq: outgoingSeq)
    )
    return [
      "type": "encrypted",
      "version": version,
      "alg": algorithm,
      "seq": outgoingSeq,
      "nonce": Data(nonce).base64URLEncodedString(),
      "ciphertext": (sealed.ciphertext + sealed.tag).base64URLEncodedString(),
    ]
  }

  func decryptPayload(_ envelope: [String: Any]) throws -> [String: Any] {
    guard let session else { throw RelayE2EError.sessionNotEstablished }
    guard
      envelope["type"] as? String == "encrypted",
      envelope["version"] as? Int == version,
      envelope["alg"] as? String == algorithm,
      let seq = envelope["seq"] as? Int,
      let nonceText = envelope["nonce"] as? String,
      let ciphertextText = envelope["ciphertext"] as? String,
      let nonceData = Data(base64URLEncoded: nonceText),
      let sealedData = Data(base64URLEncoded: ciphertextText),
      sealedData.count > 16
    else {
      throw RelayE2EError.invalidEnvelope
    }
    let ciphertext = sealedData.prefix(sealedData.count - 16)
    let tag = sealedData.suffix(16)
    let sealedBox = try ChaChaPoly.SealedBox(
      nonce: ChaChaPoly.Nonce(data: nonceData),
      ciphertext: ciphertext,
      tag: tag
    )
    let plaintext = try ChaChaPoly.open(
      sealedBox,
      using: session.key(for: incomingDirection),
      authenticating: aad(direction: incomingDirection, seq: seq)
    )
    guard let payload = try JSONSerialization.jsonObject(with: plaintext) as? [String: Any] else {
      throw RelayE2EError.invalidCiphertext
    }
    return payload
  }

  private var outgoingDirection: RelayE2EDirection {
    role == .host ? .hostToClient : .clientToHost
  }

  private var incomingDirection: RelayE2EDirection {
    role == .host ? .clientToHost : .hostToClient
  }

  private func aad(direction: RelayE2EDirection, seq: Int) -> Data {
    Data("codebuddy-remote relay e2e v1\n\(direction.rawValue)\n\(seq)".utf8)
  }
}

private struct RelayE2ESession {
  let hostToClient: SymmetricKey
  let clientToHost: SymmetricKey

  func key(for direction: RelayE2EDirection) -> SymmetricKey {
    direction == .hostToClient ? hostToClient : clientToHost
  }
}

private enum RelayE2EDirection: String {
  case hostToClient = "host-to-client"
  case clientToHost = "client-to-host"
}

private func randomBytes(count: Int) throws -> Data {
  var bytes = [UInt8](repeating: 0, count: count)
  let status = SecRandomCopyBytes(kSecRandomDefault, count, &bytes)
  guard status == errSecSuccess else {
    throw RelayE2EError.invalidCiphertext
  }
  return Data(bytes)
}

private extension Data {
  init?(base64URLEncoded value: String) {
    var base64 = value
      .replacingOccurrences(of: "-", with: "+")
      .replacingOccurrences(of: "_", with: "/")
    let padding = (4 - base64.count % 4) % 4
    base64.append(String(repeating: "=", count: padding))
    self.init(base64Encoded: base64)
  }

  func base64URLEncodedString() -> String {
    base64EncodedString()
      .replacingOccurrences(of: "+", with: "-")
      .replacingOccurrences(of: "/", with: "_")
      .replacingOccurrences(of: "=", with: "")
  }
}

File: apps/local-host/src/relay/relay-e2e.mjs
import crypto from "node:crypto";

const CURVE = "prime256v1";
const VERSION = 1;
const ALGORITHM = "P256-HKDF-SHA256-CHACHA20-POLY1305";
const NONCE_BYTES = 12;
const TAG_BYTES = 16;

export function createRelayE2EPeer({ role, pairingCode }) {
  if (role !== "host" && role !== "client") {
    throw new Error("role must be host or client");
  }
  const normalizedPairingCode = String(pairingCode || "").trim().toUpperCase();
  const ecdh = crypto.createECDH(CURVE);
  ecdh.generateKeys();

  let session;
  let outgoingSeq = 0;

  return {
    version: VERSION,
    algorithm: ALGORITHM,
    publicKey: ecdh.getPublicKey(null, "uncompressed").toString("base64url"),
    deriveSession(peerPublicKey) {
      const sharedSecret = ecdh.computeSecret(Buffer.from(String(peerPublicKey), "base64url"));
      const salt = Buffer.from(`codebuddy-remote relay e2e v1\n${normalizedPairingCode}`);
      session = {
        hostToClient: deriveKey(sharedSecret, salt, "host-to-client"),
        clientToHost: deriveKey(sharedSecret, salt, "client-to-host"),
      };
    },
    encryptPayload(payload) {
      assertSession(session);
      outgoingSeq += 1;
      const direction = outgoingDirection(role);
      const key = keyForDirection(session, direction);
      const nonce = crypto.randomBytes(NONCE_BYTES);
      const aad = aadFor(direction, outgoingSeq);
      const cipher = crypto.createCipheriv("chacha20-poly1305", key, nonce, {
        authTagLength: TAG_BYTES,
      });
      cipher.setAAD(aad);
      const plaintext = Buffer.from(JSON.stringify(payload), "utf8");
      const encrypted = Buffer.concat([cipher.update(plaintext), cipher.final()]);
      const tag = cipher.getAuthTag();
      return {
        type: "encrypted",
        version: VERSION,
        alg: ALGORITHM,
        seq: outgoingSeq,
        nonce: nonce.toString("base64url"),
        ciphertext: Buffer.concat([encrypted, tag]).toString("base64url"),
      };
    },
    decryptPayload(envelope) {
      assertSession(session);
      assertEnvelope(envelope);
      const direction = incomingDirection(role);
      const key = keyForDirection(session, direction);
      const nonce = Buffer.from(envelope.nonce, "base64url");
      const sealed = Buffer.from(envelope.ciphertext, "base64url");
      if (sealed.length <= TAG_BYTES) throw new Error("encrypted payload is too short");
      const encrypted = sealed.subarray(0, sealed.length - TAG_BYTES);
      const tag = sealed.subarray(sealed.length - TAG_BYTES);
      const decipher = crypto.createDecipheriv("chacha20-poly1305", key, nonce, {
        authTagLength: TAG_BYTES,
      });
      decipher.setAAD(aadFor(direction, envelope.seq));
      decipher.setAuthTag(tag);
      const plaintext = Buffer.concat([decipher.update(encrypted), decipher.final()]);
      return JSON.parse(plaintext.toString("utf8"));
    },
  };
}

export function isRelayEncryptedPayload(payload) {
  return Boolean(payload && typeof payload === "object" && payload.type === "encrypted");
}

function deriveKey(sharedSecret, salt, direction) {
  return Buffer.from(
    crypto.hkdfSync("sha256", sharedSecret, salt, Buffer.from(direction), 32)
  );
}

function outgoingDirection(role) {
  return role === "host" ? "host-to-client" : "client-to-host";
}

function incomingDirection(role) {
  return role === "host" ? "client-to-host" : "host-to-client";
}

function keyForDirection(session, direction) {
  return direction === "host-to-client" ? session.hostToClient : session.clientToHost;
}

function aadFor(direction, seq) {
  return Buffer.from(`codebuddy-remote relay e2e v1\n${direction}\n${seq}`);
}

function assertSession(session) {
  if (!session) throw new Error("relay e2e session is not established");
}

function assertEnvelope(envelope) {
  if (!isRelayEncryptedPayload(envelope)) throw new Error("payload is not encrypted");
  if (envelope.version !== VERSION) throw new Error("unsupported encrypted payload version");
  if (envelope.alg !== ALGORITHM) throw new Error("unsupported encrypted payload algorithm");
  if (!Number.isSafeInteger(envelope.seq) || envelope.seq < 1) {
    throw new Error("encrypted payload seq is required");
  }
  if (typeof envelope.nonce !== "string" || !envelope.nonce) {
    throw new Error("encrypted payload nonce is required");
  }
  if (typeof envelope.ciphertext !== "string" || !envelope.ciphertext) {
    throw new Error("encrypted payload ciphertext is required");
  }
}
